src/Xrefcheck/Core.hs

```diff
@@ -77,8 +74,57 @@ data Reference = Reference
   , rAnchor :: Maybe Text
     -- ^ Section or custom anchor tag.
   , rPos    :: Position
+    -- ^ Position in source file.
+  , rInfo   :: ReferenceInfo
+    -- ^ More info about the link.
   } deriving stock (Show, Generic)
 
+-- | Info about the reference.
+data ReferenceInfo
+  = RIExternal
+    -- ^ Reference to a file at outer site, e.g @[d](http://www.google.com/doodles)@
+  | RIOtherProtocol
+    -- ^ Entry not to be processed, e.g. @mailto:e-mail@
+  | RIFileLocal
+    -- ^ Reference to this file, e.g. @[a](#header)@
+  | RIFileAbsolute
+    -- ^ Reference to a file absolute to the root, e.g. @[c](/folder/file#header)@
+  | RIFileRelative
+    -- ^ Reference to a file relative to given one, e.g. @[b](folder/file#header)@
+  deriving stock (Show, Generic)
+
+-- | Compute the 'ReferenceInfo' corresponding to a given link.
+referenceInfo :: Text -> ReferenceInfo
+referenceInfo link = case toString link of
+  [] -> RIFileLocal
+  c : _             | isPathSeparator c -> RIFileAbsolute
+  '.' : c : _       | isPathSeparator c -> RIFileRelative
+  '.' : '.' : c : _ | isPathSeparator c -> RIFileRelative
+  _ | hasUrlProtocol -> RIExternal
+    | hasProtocol -> RIOtherProtocol
+    | otherwise -> RIFileRelative
+  where
+    hasUrlProtocol = "://" `T.isInfixOf` T.take 10 link
+    hasProtocol = ":"   `T.isInfixOf` T.take 10 link
+
+-- | Whether this is a link to external resource.
+isExternal :: ReferenceInfo -> Bool
+isExternal = \case
+  RIFileLocal -> False
+  RIFileRelative -> False
+  RIFileAbsolute -> False
+  RIExternal -> True
+  RIOtherProtocol -> False
```

Comment:
Could you please keep only `RIExternal` pattern and use wildcard for others?

Reply:
Yeah! Unless you prefer the other style for any reason, I prefer to not use a wildcard in this cases to be forced to specify the value when adding a new constructor.